cdap-ui/app/cdap/components/AbstractWidget/FunctionDropdownArgumentsWidget/FunctionDropdownArgumentsRow.tsx

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import * as React from 'react';
+import TextBox from 'components/AbstractWidget/FormInputs/TextBox';
+import Checkbox from '@material-ui/core/Checkbox';
+import FormControlLabel from '@material-ui/core/FormControlLabel';
+import withStyles, { StyleRules } from '@material-ui/core/styles/withStyles';
+import Select from 'components/AbstractWidget/FormInputs/Select';
+import MenuItem from '@material-ui/core/MenuItem';
+import AbstractRow, {
+  IAbstractRowProps,
+  AbstractRowStyles,
+} from 'components/AbstractWidget/AbstractMultiRowWidget/AbstractRow';
+import { parse, serialize } from 'components/AbstractWidget/FunctionDropdownArgumentsWidget/parser';
+import InputFieldDropdown from 'components/AbstractWidget/InputFieldDropdown';
+import If from 'components/If';
+import { IWidgetExtraConfig } from 'components/AbstractWidget';
+
+const styles = (theme): StyleRules => {
+  return {
+    ...AbstractRowStyles(theme),
+    inputContainer: {
+      display: 'grid',
+      gridTemplateColumns: '1fr 1fr 1fr 1fr 30px 1fr',
+      gridGap: '10px',
+    },
+    disabled: {
+      color: `${theme.palette.grey['50']}`,
+    },
+    // Align with surrounding controls
+    separator: {
+      paddingTop: '10px',
+      textAlign: 'center',
+    },
+  };
+};
+
+interface IComplexDropdown {
+  value: string | number;
+  label: string;
+}
+
+export type IDropdownOption = string | number | IComplexDropdown;
+
+interface IFunctionDropdownArgumentsRowProps extends IAbstractRowProps<typeof styles> {
+  placeholders: Record<string, string>;
+  dropdownOptions: IDropdownOption[];
+  extraConfig?: IWidgetExtraConfig;
+}
+
+interface IKeyValueState {
+  field: string;
+  func: string;
+  alias: string;
+  arguments: string;
+  ignoreNulls: boolean;
+}
+
+type StateKeys = keyof IKeyValueState;
```

Comment:
nit: We right now manually type in the keys of `IKeyValueState` in different places (say `field` or `ignoreNull`). Can we have the keys as enum and use that in both `IkeyValueState` and in `handleWidgetChange` function invocation?

Reply:
Based on a quick look just now - I don't think so. `keyof` resolves to a string literal union, which is not referenceable at run time.